src/compiler/core.ts

```diff
@@ -505,17 +505,17 @@ export function mapDefined<T, U>(array: readonly T[] | undefined, mapFn: (x: T,
 }
 
 /** @internal */
-export function* mapDefinedIterator<T, U>(iter: Iterable<T>, mapFn: (x: T) => U | undefined) {
+export function* mapDefinedIterator<T, U>(iter: Iterable<T>, mapFn: (x: T) => U | undefined): Generator<NonNullable<U>, void, unknown> {

```

Comment:
@rbuckton 
I had to also add a null check for the fixed nullchecks behavior

Reply:
Instead of `NonNullable<U>` you can use `U & ({} | null)` and you wouldn't need to change the check on L511. It's not as pretty but it matches the type of `value`:

```suggestion
export function* mapDefinedIterator<T, U>(iter: Iterable<T>, mapFn: (x: T) => U | undefined): Generator<U & ({} | null), void, unknown> {
```

Reply:
Honestly I think we should just be writing `T extends {}, U extends {}` for these.

Reply:
Hmm, it seems reasonable to me that we could have iterators through values that could be not defined.